src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,216 @@
+        }
+        size++;
+    }
+
+    public void linkFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            head.prev = newNode;
+            head = newNode;
+        }
+        size++;
+    }
+
+    public void linkByIndex(T value, int index) {
+        indexCheck(index);
+
+        if (index == 0) {
+            linkFirst(value);
+        } else if (index == size) {
+            linkLast(value);
+        } else {
+            Node<T> nodeByIndex = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            nodeByIndex.prev.next = newNode;
+            nodeByIndex.prev = newNode;
+            size++;
+        }
+    }
+
+    public Node<T> findNodeByIndex(int index) {
+        indexCheck(index);
+        Node<T> currentNode;
+
+        if (index < size >> 1) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    public void indexCheck(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(
+                    "indexCheckException: index: " + index
+                            + " size: " + size);
+        }
+    }
+
+    public void indexEqualsSizeCheck(int index) {
+        if (index == size) {
+            throw new IndexOutOfBoundsException("Can`t get element when index == size");
+        }
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        indexCheck(index);
+
+        if (index == 0) {
+            linkFirst(value);
+            return;
+        }
+
+        if (index == size) {
+            linkLast(value);
+            return;
+        }
+
+        linkByIndex(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        indexEqualsSizeCheck(index);
+
+        Node<T> x = findNodeByIndex(index);
+        T oldValue = x.value;
+        return oldValue;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexCheck(index);
+        indexEqualsSizeCheck(index);
+        Node<T> x = findNodeByIndex(index);
+        T oldValue = x.value;
+        x.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        indexEqualsSizeCheck(index);
```

Comment:
The `remove` method should not call `indexEqualsSizeCheck`, as removing an element at the last valid index is a valid operation. Consider removing this check.